### PR TITLE
feat: predictive memory scorer, knowledge architecture, and agent feedback

### PR DESCRIPTION
**Pull request by NicholaiVogel**
## The Short Version

This PR implements the intelligence layer that moves Signet from
heuristic memory injection toward learned prediction. 19 commits,
98 files, ~20K lines. The systems are co-dependent — the knowledge
architecture provides structural candidates, the predictor learns to
rank them, and agent feedback provides the training signal — so they
ship together.

This is a large change and we expect rough edges. Feedback is welcome
and wanted. We're learning as much building this as we hope our agents
will learn using it.

---

## Why This Matters

From [VISION.md](./VISION.md):

> Signet doesn't just remember — it learns what to remember. A neural
> network unique to each user, trained on their own interaction patterns,
> that gets sharper the longer you use it. No cloud. No shared weights.
> Your patterns, running locally, earning its influence by proving it
> outperforms the alternative.

This PR is a first step toward that vision. Before this, Signet
answered "what does the agent need to remember?" with a decay formula
and keyword matching. After this, the infrastructure exists for a
cross-attention neural network trained on what actually helped — as
reported by the agent itself, on every prompt, accumulated across
sessions.

"Infrastructure exists" is the honest framing. The predictor sidecar
isn't shipping a trained model — it ships cold, earns its influence
through comparison against the baseline, and only gains control as it
proves its rankings are better. The knowledge graph is structural
scaffolding that will take real usage to populate and validate. The
agent feedback mechanism is plumbed but needs connector-side work to
flow end-to-end. None of this works overnight. It's the foundation
for something that gets better over time, and we're at the beginning
of that curve.

From the [knowledge architecture](./docs/KNOWLEDGE-ARCHITECTURE.md):

> The difference between a memory system and an intelligent one is
> structure. Any system can store facts. The question is whether those
> facts are organized in a way that makes them useful — findable at the
> right moment, connected to the right things, and trimmed of everything
> that doesn't serve the goal.

---

### Knowledge Architecture (KA-1 through KA-6)

Entities, aspects, attributes, constraints, and dependencies — the
structural graph that organizes flat memories into something the agent
can walk instead of search.

- **KA-1**: Schema, types, graph helpers. Tables: `entity_aspects`,
  `entity_attributes`, `entity_dependencies`, `task_meta`.
- **KA-2**: Two-pass structural assignment pipeline (classify then
  extract dependencies).
- **KA-3**: Traversal retrieval wiring. Session-start identifies active
  entities and walks their graphs deterministically.
- **KA-5**: Continuity scoring and dashboard integration.
- **KA-6**: Feedback loop. Weight override as exploration. Entity
  pinning. Per-entity predictor win rates.

This is scaffolding. The graph needs real sessions to populate, and
the traversal needs real-world validation to prove it outperforms
pure embedding search for candidate selection.

### Predictive Memory Scorer (Sprints 1-4)

The learned ranking layer.

- **Sprint 1+2**: Cross-attention network (~1.11M params), HashTrick
  tokenizer (16K buckets), ListNet loss. JSON-RPC 2.0 sidecar client.
  RRF blending with legacy scores. Cold start ramp (alpha 0→1 over 10
  successful sessions). Topic diversity via MMR. Exploration scoring.
- **Sprint 3**: Session-end NDCG@10 comparison between predictor and
  legacy. EMA success rate. Drift detection. Training trigger.
- **Sprint 4**: Observability. Diagnostics health domain, analytics,
  dashboard PredictorTab with cold start progress and comparison history.

The predictor starts at alpha=0 (pure legacy) and only gains control
by proving its rankings outperform the baseline through NDCG comparison.
If it drifts, alpha decays back. This is deliberate — we don't trust
it until it earns trust by being right.

### Agent Relevance Feedback

The training signal.

On every `userPromptSubmit`, the daemon asks the agent which injected
memories actually helped. Scores from -1 (harmful) to 1 (directly
shaped response). Running mean accumulated across the session. Becomes
the primary training label when available.

- `<memory-feedback>` block injected into hook cycle
- Running mean in `session_memories.agent_relevance_score`
- Combined label: agent feedback (0.7) + FTS overlap (0.2) +
  continuity modifier (0.1)
- Fail-open: missing feedback falls back to existing label path

This is the piece we're most uncertain about. The theory is sound —
first-order ground truth from the agent that used the memory is better
than a model judging another model's behavior. But we haven't validated
the signal quality at scale yet, and the connector-side work to support
`memory_feedback` in hook responses isn't in this PR.

### Training Telemetry Pipeline

Foundation for eventual federated learning.

- `predictor_training_pairs` table: anonymized feature vectors +
  combined labels. No content, no prompts, no PII.
- Collected at session end in summary worker.
- Export via `/api/telemetry/training-export` (NDJSON/CSV).
- 90-day retention purge.

The long-term vision is that opted-in users contribute anonymized
pairs to a shared base model that ships with new installs, eliminating
cold start. That's far out. For now, this just collects the data
locally so we can study what good training signal looks like.

### Procedural Memory P1

Skills as graph nodes. `skill_meta` table, enrichment pipeline,
reconciler. Early stage.

---

## Known Gaps and Open Questions

We're putting this up knowing it's not perfect. Some things we're
aware of:

- **27 pre-existing test failures** — hooks tests need fixture updates
  for `getDbAccessor`, maintenance-worker tests need the new
  `predictorPipeline` config shape. These pre-date this PR but should
  be fixed.
- **Connector support for memory_feedback** — the daemon parses it,
  but no connector currently sends it. Claude Code connector work is
  next.
- **Telemetry opt-in UX** — the spec describes a conversational prompt
  and CLI toggle (`signet telemetry on/off`). Not implemented yet.
- **Rust sidecar training** — `trainFromDb` needs to know about the
  new `predictor_training_pairs` table.
- **Real-world validation** — the theory tests pass, the math checks
  out, the build is clean. But none of this has been tested against
  real user sessions at any meaningful scale.

We're open to feedback on all of it. This is a process, not a
destination.

---

## Testing

Tests are written as behavioral contracts — they encode what must be
true according to the spec, not how it's currently implemented. When
this eventually gets rewritten in a systems language, the test suite
is what the rewrite targets.

- **44 theory tests** (`tests/theory/`) — RRF fusion, cold start,
  alpha ramp, diversity, exploration, NDCG, EMA, confidence gating,
  fail-open, math invariants.
- **28 LLM integration tests** (`tests/integration/`) — statistical
  assertions against qwen3:4b via Ollama. Non-deterministic by nature,
  2/3 must pass. Graceful skip when Ollama unavailable.
- **29 session-memories tests** — feedback parsing, running mean
  accumulation.
- **20 training-pairs tests** — label computation, collection, export,
  retention purge.

## Migrations

9 new (018-026): skill_meta, knowledge structure, predictor comparisons,
checkpoint structural, entity pinning, predictor columns, predictor
comparison columns, agent feedback, predictor training pairs.

## Stats

```
98 files changed, ~20,158 insertions(+), 507 deletions(-)
692 tests passing, 27 pre-existing failures (unrelated)
```

## Test Plan

- [x] `bun test` — 692 pass, 27 pre-existing
- [x] Theory tests: 44 pass
- [x] Migration tests: fresh DB, 26 migrations, idempotent
- [x] Build passes on rebased branch
- [ ] LLM integration: `bun test ./tests/integration/` (requires Ollama)
- [ ] Manual: predictor tab renders on dashboard
- [ ] Manual: agent feedback flows through hook cycle end-to-end
- [ ] Fix pre-existing test failures (hooks fixtures, maintenance config)

---

*Built by Nicholai and Mr. Claude on the `total-recall` branch.
March 2026.*

## Summary by CodeRabbit

* **New Features**
  * Knowledge Graph integration for structured memory with entities, aspects, and dependencies.
  * Graph-based traversal for intelligent memory retrieval.
  * Predictive memory scorer for ranking relevant memories.
  * Procedural memory (skills) management with automatic enrichment.
  * Memory feedback tools to rate and improve relevance.
  * Entity pinning to prioritize knowledge in retrieval.
  * New Knowledge and Predictor tabs in dashboard.

* **Improvements**
  * Enhanced session recall with behavioral feedback loops.
  * Smarter memory injection guided by traversal paths.
  * Local-only training with federated learning support.
  * Multi-agent data isolation and privacy controls.

### CODE REVIEW

**Comment by NicholaiVogel** [comment]
@coderabbitai review
